File: apps/waddle_controller/src/util/qrOverlayPayload.ts
export const QR_OVERLAY_TEMPLATE_CUSTOM = 'custom';
export const QR_OVERLAY_TEMPLATE_HTTP = 'http';
export const QR_OVERLAY_TEMPLATE_MAILTO = 'mailto';
export const QR_OVERLAY_TEMPLATE_TEL = 'tel';
export const QR_OVERLAY_TEMPLATE_SMS = 'sms';
export const QR_OVERLAY_TEMPLATE_GEO = 'geo';
export const QR_OVERLAY_TEMPLATE_WIFI = 'wifi';
export const QR_OVERLAY_TEMPLATE_VCARD = 'vcard';
export const QR_OVERLAY_TEMPLATE_VCALENDAR = 'vcalendar';

export const QR_OVERLAY_TEMPLATE_VALUES = [
  QR_OVERLAY_TEMPLATE_CUSTOM,
  QR_OVERLAY_TEMPLATE_HTTP,
  QR_OVERLAY_TEMPLATE_MAILTO,
  QR_OVERLAY_TEMPLATE_TEL,
  QR_OVERLAY_TEMPLATE_SMS,
  QR_OVERLAY_TEMPLATE_GEO,
  QR_OVERLAY_TEMPLATE_WIFI,
  QR_OVERLAY_TEMPLATE_VCARD,
  QR_OVERLAY_TEMPLATE_VCALENDAR,
] as const;

export type QrOverlayTemplate = (typeof QR_OVERLAY_TEMPLATE_VALUES)[number];

export const QR_OVERLAY_TEMPLATE_LABELS: Record<QrOverlayTemplate, string> = {
  custom: 'Custom (raw data)',
  http: 'URL (http/https)',
  mailto: 'Email (mailto)',
  tel: 'Phone (tel)',
  sms: 'SMS (smsto)',
  geo: 'Location (geo)',
  wifi: 'Wi‑Fi (WIFI:…)',
  vcard: 'Contact (vCard)',
  vcalendar: 'Event (vCalendar)',
};

function readString(raw: unknown): string {
  if (raw == null) return '';
  if (typeof raw === 'string') return raw.trim();
  return String(raw).trim();
}

function escapeWifiField(value: string): string {
  let out = '';
  for (let i = 0; i < value.length; i++) {
    const c = value[i];
    if (c === '\\' || c === ';' || c === ',' || c === ':') {
      out += '\\';
    }
    out += c;
  }
  return out;
}

function normalizePhone(raw: string): string {
  const trimmed = raw.replace(/\s+/g, '');
  if (!trimmed) return '';
  if (trimmed.startsWith('+')) return trimmed;
  return trimmed;
}

export function buildQrOverlayPayload(
  template: string,
  fields: Record<string, unknown>,
): string {
  const t = template.trim().toLowerCase();
  switch (t) {
    case QR_OVERLAY_TEMPLATE_CUSTOM:
      return readString(fields.payload ?? fields.data);
    case QR_OVERLAY_TEMPLATE_HTTP:
      return buildHttpPayload(fields);
    case QR_OVERLAY_TEMPLATE_MAILTO:
      return buildMailtoPayload(fields);
    case QR_OVERLAY_TEMPLATE_TEL:
      return buildTelPayload(fields);
    case QR_OVERLAY_TEMPLATE_SMS:
      return buildSmsPayload(fields);
    case QR_OVERLAY_TEMPLATE_GEO:
      return buildGeoPayload(fields);
    case QR_OVERLAY_TEMPLATE_WIFI:
      return buildWifiPayload(fields);
    case QR_OVERLAY_TEMPLATE_VCARD:
      return buildVcardPayload(fields);
    case QR_OVERLAY_TEMPLATE_VCALENDAR:
      return buildVcalendarPayload(fields);
    default:
      return '';
  }
}

function buildHttpPayload(fields: Record<string, unknown>): string {
  let url = readString(fields.url);
  if (!url) return '';
  if (!url.includes('://')) {
    url = `https://${url}`;
  }
  return url;
}

function buildMailtoPayload(fields: Record<string, unknown>): string {
  const email = readString(fields.email);
  if (!email) return '';
  const subject = readString(fields.subject);
  const body = readString(fields.body);
  const params: string[] = [];
  if (subject) params.push(`subject=${encodeURIComponent(subject)}`);
  if (body) params.push(`body=${encodeURIComponent(body)}`);
  return params.length > 0 ? `mailto:${email}?${params.join('&')}` : `mailto:${email}`;
}

function buildTelPayload(fields: Record<string, unknown>): string {
  const phone = normalizePhone(readString(fields.phone));
  return phone ? `tel:${phone}` : '';
}

function buildSmsPayload(fields: Record<string, unknown>): string {
  const phone = normalizePhone(readString(fields.phone));
  if (!phone) return '';
  const body = readString(fields.body);
  return body ? `smsto:${phone}?body=${encodeURIComponent(body)}` : `smsto:${phone}`;
}

function buildGeoPayload(fields: Record<string, unknown>): string {
  const lat = readString(fields.lat);
  const lng = readString(fields.lng);
  if (!lat || !lng) return '';
  const label = readString(fields.label);
  return label ? `geo:${lat},${lng}?q=${encodeURIComponent(label)}` : `geo:${lat},${lng}`;
}

function buildWifiPayload(fields: Record<string, unknown>): string {
  const ssid = readString(fields.ssid);
  if (!ssid) return '';
  const security = readString(fields.securityType);
  const t = security || 'nopass';
  const password = readString(fields.password);
  const hidden = fields.hidden === true;
  let out = `WIFI:T:${escapeWifiField(t)};S:${escapeWifiField(ssid)};`;
  if (t !== 'nopass' && t !== 'NOPASS' && password) {
    out += `P:${escapeWifiField(password)};`;
  }
  if (hidden) {
    out += 'H:true;';
  }
  return out;
}

function buildVcardPayload(fields: Record<string, unknown>): string {
  const firstName = readString(fields.firstName);
  const lastName = readString(fields.lastName);
  const fullName = readString(fields.fullName);
  const fn =
    fullName || [firstName, lastName].filter((s) => s.length > 0).join(' ').trim();
  if (!fn) return '';
  const lines = ['BEGIN:VCARD', 'VERSION:3.0', `FN:${fn}`];
  const org = readString(fields.org);
  if (org) lines.push(`ORG:${org}`);
  const phone = readString(fields.phone);
  if (phone) lines.push(`TEL:${phone}`);
  const email = readString(fields.email);
  if (email) lines.push(`EMAIL:${email}`);
  const title = readString(fields.title);
  if (title) lines.push(`TITLE:${title}`);
  lines.push('END:VCARD');
  return lines.join('\n');
}

function buildVcalendarPayload(fields: Record<string, unknown>): string {
  const summary = readString(fields.summary);
  const dtStart = readString(fields.dtStart);
  const dtEnd = readString(fields.dtEnd);
  if (!summary || !dtStart || !dtEnd) return '';
  const lines = [
    'BEGIN:VCALENDAR',
    'VERSION:2.0',
    'BEGIN:VEVENT',
    `SUMMARY:${summary}`,
    `DTSTART:${dtStart}`,
    `DTEND:${dtEnd}`,
  ];
  const location = readString(fields.location);
  if (location) lines.push(`LOCATION:${location}`);
  const description = readString(fields.description);
  if (description) lines.push(`DESCRIPTION:${description}`);
  lines.push('END:VEVENT', 'END:VCALENDAR');
  return lines.join('\n');
}

export function readQrOverlayTemplate(raw: Record<string, unknown>): QrOverlayTemplate {
  const t = readString(raw.template).toLowerCase();
  if ((QR_OVERLAY_TEMPLATE_VALUES as readonly string[]).includes(t)) {
    return t as QrOverlayTemplate;
  }
  return QR_OVERLAY_TEMPLATE_CUSTOM;
}

export function readTemplateFields(raw: Record<string, unknown>): Record<string, unknown> {
  const f = raw.template_fields;
  if (f != null && typeof f === 'object' && !Array.isArray(f)) {
    return { ...(f as Record<string, unknown>) };
  }
  return {};
}

export function syncQrOverlayFormData(formData: Record<string, unknown>): Record<string, unknown> {
  const template = readQrOverlayTemplate(formData);
  const templateFields = readTemplateFields(formData);
  let payload = buildQrOverlayPayload(template, templateFields);
  if (!payload && template === QR_OVERLAY_TEMPLATE_CUSTOM) {
    payload = readString(formData.payload);
  }
  return {
    ...formData,
    template,
    template_fields: templateFields,
    payload,
  };
}

File: packages/waddle_shared/lib/persistence/qr_overlay_payload.dart
/// Built-in QR overlay payload templates (operator-facing).
const String kQrOverlayTemplateCustom = 'custom';
const String kQrOverlayTemplateHttp = 'http';
const String kQrOverlayTemplateMailto = 'mailto';
const String kQrOverlayTemplateTel = 'tel';
const String kQrOverlayTemplateSms = 'sms';
const String kQrOverlayTemplateGeo = 'geo';
const String kQrOverlayTemplateWifi = 'wifi';
const String kQrOverlayTemplateVcard = 'vcard';
const String kQrOverlayTemplateVcalendar = 'vcalendar';

/// All supported template ids for schema and validation.
const List<String> kQrOverlayTemplateValues = [
  kQrOverlayTemplateCustom,
  kQrOverlayTemplateHttp,
  kQrOverlayTemplateMailto,
  kQrOverlayTemplateTel,
  kQrOverlayTemplateSms,
  kQrOverlayTemplateGeo,
  kQrOverlayTemplateWifi,
  kQrOverlayTemplateVcard,
  kQrOverlayTemplateVcalendar,
];

/// Builds the QR payload string from [template] and [fields].
/// Returns empty string when required inputs are missing or invalid.
String buildQrOverlayPayload(String template, Map<String, dynamic> fields) {
  final t = template.trim().toLowerCase();
  switch (t) {
    case kQrOverlayTemplateCustom:
      return _readString(fields['payload'] ?? fields['data']).trim();
    case kQrOverlayTemplateHttp:
      return _buildHttpPayload(fields);
    case kQrOverlayTemplateMailto:
      return _buildMailtoPayload(fields);
    case kQrOverlayTemplateTel:
      return _buildTelPayload(fields);
    case kQrOverlayTemplateSms:
      return _buildSmsPayload(fields);
    case kQrOverlayTemplateGeo:
      return _buildGeoPayload(fields);
    case kQrOverlayTemplateWifi:
      return _buildWifiPayload(fields);
    case kQrOverlayTemplateVcard:
      return _buildVcardPayload(fields);
    case kQrOverlayTemplateVcalendar:
      return _buildVcalendarPayload(fields);
    default:
      return '';
  }
}

String _buildHttpPayload(Map<String, dynamic> fields) {
  var url = _readString(fields['url']);
  if (url.isEmpty) {
    return '';
  }
  if (!url.contains('://')) {
    url = 'https://$url';
  }
  return url;
}

String _buildMailtoPayload(Map<String, dynamic> fields) {
  final email = _readString(fields['email']);
  if (email.isEmpty) {
    return '';
  }
  final subject = _readString(fields['subject']);
  final body = _readString(fields['body']);
  final buf = StringBuffer('mailto:$email');
  final params = <String>[];
  if (subject.isNotEmpty) {
    params.add('subject=${Uri.encodeComponent(subject)}');
  }
  if (body.isNotEmpty) {
    params.add('body=${Uri.encodeComponent(body)}');
  }
  if (params.isNotEmpty) {
    buf.write('?${params.join('&')}');
  }
  return buf.toString();
}

String _buildTelPayload(Map<String, dynamic> fields) {
  final phone = _normalizePhone(_readString(fields['phone']));
  if (phone.isEmpty) {
    return '';
  }
  return 'tel:$phone';
}

String _buildSmsPayload(Map<String, dynamic> fields) {
  final phone = _normalizePhone(_readString(fields['phone']));
  if (phone.isEmpty) {
    return '';
  }
  final body = _readString(fields['body']);
  if (body.isEmpty) {
    return 'smsto:$phone';
  }
  return 'smsto:$phone?body=${Uri.encodeComponent(body)}';
}

String _buildGeoPayload(Map<String, dynamic> fields) {
  final lat = _readString(fields['lat']);
  final lng = _readString(fields['lng']);
  if (lat.isEmpty || lng.isEmpty) {
    return '';
  }
  final label = _readString(fields['label']);
  if (label.isEmpty) {
    return 'geo:$lat,$lng';
  }
  return 'geo:$lat,$lng?q=${Uri.encodeComponent(label)}';
}

String _buildWifiPayload(Map<String, dynamic> fields) {
  final ssid = _readString(fields['ssid']);
  if (ssid.isEmpty) {
    return '';
  }
  final security = _readString(fields['securityType']);
  final t = security.isEmpty ? 'nopass' : security;
  final password = _readString(fields['password']);
  final hidden = fields['hidden'] == true;
  final buf = StringBuffer('WIFI:T:${_escapeWifiField(t)};');
  buf.write('S:${_escapeWifiField(ssid)};');
  if (t != 'nopass' && t != 'NOPASS' && password.isNotEmpty) {
    buf.write('P:${_escapeWifiField(password)};');
  }
  if (hidden) {
    buf.write('H:true;');
  }
  return buf.toString();
}

String _buildVcardPayload(Map<String, dynamic> fields) {
  final firstName = _readString(fields['firstName']);
  final lastName = _readString(fields['lastName']);
  final fullName = _readString(fields['fullName']);
  final fn = fullName.isNotEmpty
      ? fullName
      : [firstName, lastName].where((s) => s.isNotEmpty).join(' ').trim();
  if (fn.isEmpty) {
    return '';
  }
  final lines = <String>[
    'BEGIN:VCARD',
    'VERSION:3.0',
    'FN:$fn',
  ];
  final org = _readString(fields['org']);
  if (org.isNotEmpty) {
    lines.add('ORG:$org');
  }
  final phone = _readString(fields['phone']);
  if (phone.isNotEmpty) {
    lines.add('TEL:$phone');
  }
  final email = _readString(fields['email']);
  if (email.isNotEmpty) {
    lines.add('EMAIL:$email');
  }
  final title = _readString(fields['title']);
  if (title.isNotEmpty) {
    lines.add('TITLE:$title');
  }
  lines.add('END:VCARD');
  return lines.join('\n');
}

String _buildVcalendarPayload(Map<String, dynamic> fields) {
  final summary = _readString(fields['summary']);
  final dtStart = _readString(fields['dtStart']);
  final dtEnd = _readString(fields['dtEnd']);
  if (summary.isEmpty || dtStart.isEmpty || dtEnd.isEmpty) {
    return '';
  }
  final lines = <String>[
    'BEGIN:VCALENDAR',
    'VERSION:2.0',
    'BEGIN:VEVENT',
    'SUMMARY:$summary',
    'DTSTART:$dtStart',
    'DTEND:$dtEnd',
  ];
  final location = _readString(fields['location']);
  if (location.isNotEmpty) {
    lines.add('LOCATION:$location');
  }
  final description = _readString(fields['description']);
  if (description.isNotEmpty) {
    lines.add('DESCRIPTION:$description');
  }
  lines
    ..add('END:VEVENT')
    ..add('END:VCALENDAR');
  return lines.join('\n');
}

String _escapeWifiField(String value) {
  final buf = StringBuffer();
  for (var i = 0; i < value.length; i++) {
    final c = value[i];
    if (c == r'\' || c == ';' || c == ',' || c == ':') {
      buf.write(r'\');
    }
    buf.write(c);
  }
  return buf.toString();
}

String _readString(Object? raw) {
  if (raw == null) {
    return '';
  }
  if (raw is String) {
    return raw.trim();
  }
  return raw.toString().trim();
}

String _normalizePhone(String raw) {
  final trimmed = raw.replaceAll(RegExp(r'\s+'), '');
  if (trimmed.isEmpty) {
    return '';
  }
  if (trimmed.startsWith('+')) {
    return trimmed;
  }
  return trimmed;
}

bool isQrOverlayTemplateValid(String? template) {
  if (template == null || template.trim().isEmpty) {
    return false;
  }
  return kQrOverlayTemplateValues.contains(template.trim().toLowerCase());
}
